File: python/array/MergeIntervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    # @param intervals, a list of Interval
    # @return a list of Interval
    def merge(self, intervals):
        result = []
        if len(intervals) == 0:
            return result
            
        intervals = sorted(intervals, key = lambda i: i.start)
        
        result.append(intervals[0])
        i = 1
        while i < len(intervals):
            if intervals[i].start > result[-1].end:
                result.append(intervals[i])
            else:
                result[-1].end = max(intervals[i].end, result[-1].end)
            i += 1
            
        return result

File: cpp/array/MergeIntervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
bool comparator (Interval i, Interval j) {
    return (i.start < j.start);
}

class Solution {
public:
    std::vector<Interval> merge(std::vector<Interval> &intervals) {
        std::sort(intervals.begin(), intervals.end(), comparator);
        std::vector<Interval> result;
        if (intervals.size() == 0) {
            return result;
        }
        int i = 1;
        result.push_back(intervals[0]);
        while (i < intervals.size()) {
            if (intervals[i].start > result.back().end) {
                result.push_back(intervals[i]);
            } else {
                result.back().end = std::max(intervals[i].end, result.back().end);
            }
            i ++;
        }
        return result;
    }
};
